### ISSUE TITLE
Onboard SATA

### ISSUE THREAD

**Issue by alexarda** [fifteenhex/crappy96corearmserverexperiments#1]
Hi fifteenhex

Thanks for this, I managed to boot one of these boards with an ATX PSU by following this repo.

Did you happen to check whether your SATA ports worked directly from the board? I can't get them to work on mine, and I don't have the backplane to test if its a direct connection issue.

**Comment by fifteenhex** [comment]
Hi,

Nice. I didn't expect anyone to actually find this useful.

SATA I haven't tried. I'm using an NVMe SSD but in a USB3 adapter thing.
If I find some time over the weekend I'll try hooking up a SATA drive and see what happens.

**Comment by alexarda** [comment]
Please disregard this. It was a rookie error on my part. I attached another drive and it worked perfectly. 

It's stupid because the other drive that didn't work is known good, and I switched out both power and data cable with no result. 

But then I tried another drive and it was detected. 

Apologies for wasting your time and thanks for documenting everything!

**Comment by fifteenhex** [comment]
Nice. I wonder what the difference is? My long term plan was to work out how to get an NVMe drive to work properly in the PCIe slot.

**Comment by alexarda** [comment]
Maybe a vendor thing? The non-working drive is an Intel - quite an old one too. 

Interesting, an Optane NVMe drive also has issues. The functional vendors for me are Adata for SATA and Samsung for NVMe.

**Comment by fifteenhex** [comment]
> Maybe a vendor thing? The non-working drive is an Intel - quite an old one too.
> 
> Interesting, an Optane NVMe drive also has issues. The functional vendors for me are Adata for SATA and Samsung for NVMe.

kioxia nvme didn't work for me. It looks like it's working initially and then stops responding. So the machine boots but eventually goes crazy because it can't read/write data anymore.

I'll try to find another nvme and give it a go. I assumed that the pcie was just garbage. :)

**Comment by alexarda** [comment]
![Screenshot from 2023-05-29 06-16-55](https://github.com/fifteenhex/crappy96corearmserverexperiments/assets/31119958/602c3410-c257-42aa-a831-bf3a53485411)

If you're interested I managed to get CentOS Stream reasonably stable with an AMD GPU, which is nice.

**Comment by no2chem** [comment]
Hi there, thanks for all of this, it's been pretty useful.

I purchased a gigabyte R270-T60 on eBay hoping to use it as a NAS. Your tips helped - I'd add that arm-smmu.disable_bypass=n is needed with new kernels.

Have you been able to get normal hard drives working? I can't seem to get some Seagate exos drives working, at least with the backplane it came with.

**Comment by ayakael** [comment]
@no2chem You must've gotten that R270-T60 from the same seller as I did (erik something something). I gave up on SATA. Every one of my drivers gave me some variation of `ata1: SATA link down` or `link online but 1 devices misclassified`, so I put in an LSI HBA. Fortunately, the backplane even does SAS! That said, nvme has worked as long as I do not set `acpi=force` like I've seen in few other guides.

**Comment by no2chem** [comment]
@ayakael Actually, I figured out the SATA - the issue appears to be that the SATA ports don't support spread spectrum. WD drives work fine, but Seagate drives require disabling spread spectrum via seatools - you need to plug it into another device that supports spread spectrum and disable it. Works reliably after disabling spread spectrum.

**Comment by ayakael** [comment]
@no2chem Many thanks, I'll try that out! By the way, when you do lspci, do you see any Crypto acceleration device on your server?

**Comment by no2chem** [comment]
I don't, is there a specific accelerator you're looking for?
I see

```
lspci | grep accelerator

0000:00:09.0 Processing accelerators: Cavium, Inc. THUNDERX Random Number Generator (rev 09)
0000:00:09.1 Processing accelerators: Cavium, Inc. THUNDERX Random Number Generator virtual function (rev 09)
0000:03:00.0 Processing accelerators: Cavium, Inc. THUNDERX Zip Coprocessor (rev 09)
0000:04:00.0 Processing accelerators: Cavium, Inc. THUNDERX DFA (rev 09)
000a:00:09.0 Processing accelerators: Cavium, Inc. THUNDERX Random Number Generator (rev 09)
000a:00:09.1 Processing accelerators: Cavium, Inc. THUNDERX Random Number Generator virtual function (rev 09)
000a:03:00.0 Processing accelerators: Cavium, Inc. THUNDERX Zip Coprocessor (rev 09)
000a:04:00.0 Processing accelerators: Cavium, Inc. THUNDERX DFA (rev 09)
```

**Comment by ayakael** [comment]
RIght, some ThunderX systems have crypto accelerators, which isn't my case. I'm now trying to get an Intel QAT accelerator going, but I'm experiencing MSI-X related errors. Apparently, MSI-X is broken on my system, trying to debug what's wrong. The more I use this server, the more I realize that a bunch of stuff is just plain broken.

**Comment by keixthb** [comment]
> Hi fifteenhex
> 
> Thanks for this, I managed to boot one of these boards with an ATX PSU by following this repo.
> 
> Did you happen to check whether your SATA ports worked directly from the board? I can't get them to work on mine, and I don't have the backplane to test if its a direct connection issue.

Hey @alexarda, could you explain  how you got this to boot from a ATX PSU? I just picked up one of these devices and was surprised it did not come with the standard connectors.

**Comment by alexarda** [comment]
@keixthb yes, absolutely. So the pinout can be found in another issue thread [here](https://github.com/fifteenhex/crappy96corearmserverexperiments/issues/3 ).

I bought one of [these adapters](https://www.moddiy.com/products/24-8-Pin-to-Mini-18-Pin-Adapter-Cable-20cm-for-Dell-C6100-L5639-L5520.html) from ModDIY and repined it. 

I did just use the 5vsb from the ATX PSU rather than try and get 12vsb but it seems to work...

**Comment by keixthb** [comment]
@alexarda ok, I just ordered one, thanks so much. I'll see if i can get it to work and post an update in the next few weeks or so

**Comment by keixthb** [comment]
@alexarda How did you re wire the power cable, and did you use 2 of them on the machine? I wired mine following the 18 pin diagram and it didn't boot, but I only purchased one... I must be doing something wrong. This is where I'm at with the project:


![power_cable](https://github.com/user-attachments/assets/a835a828-167f-468c-823b-2d0b43d21906)

![node_with_fans](https://github.com/user-attachments/assets/be4a84cd-ce6e-4f5f-be53-1a33c8a10fb9)


...also, I designed a fan bracket for a 120mm if anyone wants a copy of the .stl. 




<img width="733" alt="capg_bracket" src="https://github.com/user-attachments/assets/8f84571c-871f-4e89-8f0d-a0773c91cdb8">

Thanks!

**Comment by keixthb** [comment]
@alexarda I tried connecting pin 10 (on the 18pin) to the pin 9 for the +5vsb, and the lights come on, but I don't get anything on the vga out. 

![Screenshot 2024-10-08 at 8 42 56 AM](https://github.com/user-attachments/assets/6f337694-4a58-4cf6-a657-d8bea1b7effd)

**Comment by alexarda** [comment]
Hi @keixthb, sorry I’ve been a bit lax with this. This went into storage after I bought an Ampere system. I can dig this out of storage for you this weekend. From memory serial is the best option to start tinkering. VGA works but much later in the boot process and you might catch issues on serial  you wouldn’t otherwise see

**Comment by keixthb** [comment]
@alexarda No worries. I would appreciate that, thank you!  Ideally, my goal is to install RHEL 9 on the system so I can test some of the Nvidia cards (centos is good too though). I know the tesla line is sensitive with firmware so I'd like to install the driver for a bunch and see which ones work--and which ones don't.

**Comment by alexarda** [comment]
@keixthb this is my version

![IMG_4390](https://github.com/user-attachments/assets/dbbb78b8-c93b-425b-a0ff-c2c27eb8f7a2)

It comes from 1 x 24 pin ATX and 1 x 8 pin EPS.

Pinout is all 12v and GND as in the other thread EXCEPT I use 5vstb from the ATX supply on pin 10.

Bit difficult to show broken out, but here you go:

![IMG_4387](https://github.com/user-attachments/assets/6563ec7c-9637-4178-994a-58aa6a41a8bd)

![IMG_4388](https://github.com/user-attachments/assets/541f4efc-269d-4d52-85ca-7df923ef202f)

**Comment by alexarda** [comment]
That single heat shrunk cover cable is a dumb error. It's meant to be jumper to power the ATX supply on PS_On to GND. I thought I could get the board's power button to work, so I depinned it and in so doing broke the pin. 

I never got further, but when I need to power the board it goes to pin 11 of the front panel header. 
![Screenshot from 2024-10-11 10-22-41](https://github.com/user-attachments/assets/3e841d76-b67a-4830-8901-9da83432a334)

**Comment by keixthb** [comment]
@alexarda I got it to turn on with the cable, I'll work on the serial next. How did you get the centos 9 kernel installed? Did you use a aarch64 dvd iso and boot from a USB stick?

**Comment by alexarda** [comment]
From memory, yes. This is from my notes at the time:

CentOS: Tested = 

acpi=force --> install wont display correctly, fail

acpi=force pci=noaer pcie_aspm=off --> install wont display correctly, fail

acpi=force pci=noaer pcie_aspm=off modprobe.blacklist=ast --> failed at startx after install

acpi=force pci=noaer pcie_aspm=off console=ttyS0 --> **install success!**

After install changed to = acpi=force pci=noaer pcie_aspm=off modprobe.blacklist=ast  amdgpu.dpm=0 -->  Working!

Then: systemctl set-default graphical.target

Poweroff, insert AMDGPU

**Comment by keixthb** [comment]
@alexarda Thank you! Okay, so you installed the linux kernel first--then the gpu. I ordered a serial cable for the machine that should be here today. I'll try it with `` acpi=force``  , ``pci=noaer``,  ``pcie_aspm=off``, ``modprobe.blacklist=ast`` and see what happens.

**Comment by keixthb** [comment]
So, using ``sudo minimum -D /dev/ttyUSB0``, I configure with:

![Screenshot 2024-10-13 at 2 48 38 PM](https://github.com/user-attachments/assets/a3c39dda-786a-4e1f-85e3-078725ef9828)

Which yields the following when the gigabyte server turns on:

![Screenshot 2024-10-13 at 2 43 28 PM](https://github.com/user-attachments/assets/98eb4274-ebb0-4e3a-bbd9-0e7ce00a4a55)

I downloaded the ``ARM64 (aarch64)`` version of centos stream 9 [here](https://centos.org/download/), and flash a USB drive using balena etcher:

![Screenshot 2024-10-13 at 2 57 21 PM](https://github.com/user-attachments/assets/fe89f581-da42-4eb5-b640-6baff953d68f)

I insert the usb stick into the machine as shown:

![Screenshot 2024-10-13 at 3 27 08 PM](https://github.com/user-attachments/assets/e2fc39b8-244c-4fcb-9810-bed012b99f6e)

And the result is the same on the serial:

![Screenshot 2024-10-13 at 3 25 23 PM](https://github.com/user-attachments/assets/b3437729-c36d-4ed7-9c00-6badf1872456)

**Comment by alexarda** [comment]
This might be a hardware/firmware issue. Possibly ram at that point. It should boot and drop you to an EFI shell.

I don’t think I captured a boot log. But I might get a chance to do so later this week if it would help

**Comment by keixthb** [comment]
@alexarda that would be great, thank you. I will work on it again and see if i can figure out what's going on my end. I think I have some ram sticks I can pull from another working machine to test with.

**Comment by keixthb** [comment]
@alexarda Can you send a picture of the memory you are using with your machine? I tried two separate sets and I get the same memory controller error.

**Comment by alexarda** [comment]
@keixthb boot log as mentioned:
[MT70-HD0_BootLog.txt](https://github.com/user-attachments/files/17448459/MT70-HD0_BootLog.txt)

Note the very different BDK versions. On your screenshot, it looks like it stalls at the point in my log where we can see _BMC IP: N/A_, which is immediately followed by the ram configuration/testing/training info. 

My suspicion is still a ram issue, but I wouldn't rule out some weird BMC thing too. These motherboards are a total mess... 

The part numbers of my ram modules are all M393A1G43DB0-CPB - that's in the log too but here also for posterity.

**Comment by keixthb** [comment]
@alexarda Thanks, I just ordered the memory. I'll update once it arrives.

**Comment by keixthb** [comment]
@alexarda good news, the efi works! With the memory installed in the furthest blue slots away from the CPU sockets, I am able to see the menu:

![Screenshot 2024-10-28 at 7 57 06 PM](https://github.com/user-attachments/assets/abcdd7c7-94cb-49b9-a5bd-b9c9d0371262)

Using a prebuilt rhel9 image for a raspberry pi 5 I am greeted with grub:

![Screenshot 2024-10-28 at 7 57 27 PM](https://github.com/user-attachments/assets/c6220fc4-c309-46b0-aee7-880ec5b588df)

I change the parameters with the following:


![Screenshot 2024-10-28 at 7 57 41 PM](https://github.com/user-attachments/assets/0133d527-528e-4aaf-b374-c9a6c6dbd064)

But it appears to be hanging on "booting a command list" for both rescue and the top kernel 5.14

**Comment by keixthb** [comment]
For reference, this is the content of my /dev/ttyUSB0 and /dev/ttyS0:

![Screenshot 2024-10-28 at 8 42 07 PM](https://github.com/user-attachments/assets/a19420ed-93ba-4896-b24e-7e5d6d8c2cc2)

**Comment by keixthb** [comment]
@alexarda can you explain how you constructed your boot image? I can't quite figure out the way of constructing one myself. The only way i can launch grub is using the pi image and I think something must be different enough between the architectures that's preventing it from launching properly.

**Comment by alexarda** [comment]
@keixthb did you manage to update the firmware? Later versions are still buggy but possibly less buggy. 

I'd start there. The build date in your screenshot is 3 years prior to the latest firmware build [available here](https://www.gigabyte.com/Enterprise/ARM-Server/H270-T70-rev-110#Support-Bios)

**Comment by keixthb** [comment]
@alexarda Hello, sorry for the late response. Yes. I was able to update the firmware, and that seemed to solve the problem with the rhel image not loading correctly. Basically:

 - I created a bootable image on a USB drive (using the rhel 10 aarch64 dvd .iso and balena etcher on macOS).
 -  I installed the rhel 10 image on a micro SD card using a RPI5 and booted into the linux kernel that way. 
 - Then I removed the SD card, put it into a microSD-USB adapter and inserted that into the gigabyte blade. 
 
 It took a few reboots but it eventually booted into the desktop. As far as the system goes, it seems that the gnu linker is broken... I had to manually specify flags for the linker but I did manage to get kokkos to build on it:

![results_crappy_96_core_arm_server_experiments](https://github.com/user-attachments/assets/58f9bc59-c9ee-4220-b196-65b3bc5d10e2)

With a modern conjugate gradient solver, I'm getting 0.014642/0.110242 =0.133 X the performance (GFLOP/s) of the GH200 with kokkos serial. Which isn't great... but you get 96 cores on this blade VS the 72 on the GH200. 


Other than that, so far I am quite pleased with the system. Thanks so much for your help! I'm trying to get a NVidia GTX1060 to run on it right now, but it seems to hang on the boot process and won't load the kernel. Do you have any advice? Were there any tricks that you did with your AMD card? Thanks again!!

**Comment by keixthb** [comment]
Just for completeness, this was my working hardware setup:

![hardware_setup](https://github.com/user-attachments/assets/dc9733ca-9700-4582-9801-9404377f18ce)

**Comment by alexarda** [comment]
Congratulations on getting to the point of a working setup! Yeah, as you point out performance is basically terrible…

To be honest I’m not much help with troubleshooting Nvidia issues. I’m much more familiar with AMD on Linux (and more recently Intel). I did recently set up OpenSuse Leap on an Ampere Altra system with a GTX 960, but didn’t like the experience and went back to AMD pretty quickly.
